File: ts_sdk/src/entities/spreadsheet.ts
import { APIEntity, registerEntity } from '../APIEntity';
import { DockPointerData } from '../models/DockPointer';

/**
 * Spreadsheet entity — a flat tabular file (`.csv` or `.xlsx`) backed by a
 * SpreadsheetRecord on disk, discovered anywhere in a project (like markdown).
 *
 * `asset_ref` is the file itself. The grid editor reads it: CSV is editable and
 * saved back through the plain-text FS write path; XLSX is parsed frontend-side
 * (SheetJS) and shown read-only. `format` ("csv" | "xlsx") drives that branch.
 */
@registerEntity
export class Spreadsheet extends APIEntity<Spreadsheet> {
  static type: string = 'spreadsheet';
  static override icon = 'Table';

  title: string = '';
  description: string = '';
  asset_ref?: string;
  /** "csv" | "xlsx" — the on-disk format; drives the editable-vs-readonly branch. */
  format?: string;
  num_rows?: number;
  num_cols?: number;
  /** XLSX workbook sheet names (empty for CSV). */
  sheet_names?: string[];

  constructor(entity: Partial<Spreadsheet> = {}) {
    super(entity);
    this.title = entity.title ?? '';
    this.description = entity.description ?? '';
    this.asset_ref = entity.asset_ref;
    this.format = entity.format;
    this.num_rows = entity.num_rows;
    this.num_cols = entity.num_cols;
    this.sheet_names = entity.sheet_names ?? [];
  }

  /** Default open target: the spreadsheet grid editor (URL-first target). */
  override get dockPointer(): DockPointerData {
    return this.assetEditorPointer('spreadsheet') ?? super.dockPointer;
  }

  override get editorDockPointer(): DockPointerData {
    return this.assetEditorPointer('spreadsheet') ?? super.editorDockPointer;
  }

  override get searchDockPointer(): DockPointerData {
    return this.assetEditorPointer('spreadsheet') ?? this.dockPointer;
  }
}

File: flow_sdk/builtin/spreadsheet.py
"""``Spreadsheet`` — a flat-file-backed tabular asset (CSV or XLSX).

Unlike the folder-backed ``Deck``, a spreadsheet is a single file discovered
anywhere in a project (like ``Markdown``):

    <project>/**/*.csv     — plain text, editable
    <project>/**/*.xlsx    — binary (Office Open XML), viewed read-only

The file itself is the entity; the frontend grid (RevoGrid) reads/edits the
bytes. CSV round-trips through the plain text FS write path; XLSX is parsed
frontend-side (SheetJS) and shown read-only.

The walker + extractor + id-mint live in
``flow_sdk/fs_store/indexer/functions/spreadsheet.py``; the type registration
lives in ``flow_sdk/schema/type_info/spreadsheet_type_info.py``. Modeled on the
MARKDOWN (flat-file) indexer, not the DECK (folder-marker) one.
"""
from __future__ import annotations

from typing import List, Optional

from flow_sdk.api.api_types.api_field import APIField, Sharing
from flow_sdk.core import Entity


class Spreadsheet(Entity):
    type: str = APIField(default="spreadsheet")
    title: str = APIField("")
    description: Optional[str] = APIField(None, blob=True)

    # "csv" | "xlsx" — the on-disk format, denormalized by the extractor from the
    # file suffix. Drives the frontend editor branch (editable grid vs read-only).
    format: str = APIField("")
    num_rows: int = APIField(0)
    num_cols: int = APIField(0)
    # XLSX only — the workbook's sheet names (empty for CSV, a single logical sheet).
    sheet_names: List[str] = APIField(default_factory=list)

    # Absolute path of the backing file on disk, stamped by the indexer /
    # ``Entity.from_fs_ref``. A plain string, mirrors MARKDOWN.
    asset_ref: str = APIField(default="", sharing=Sharing.PRIVATE)
